File: ios/Shared/WordPredictionEngine.swift
import Foundation

/**
 * WordPredictionEngine - Loads and queries word prediction data
 * 
 * Uses binary format with direct memory access for efficiency.
 * Supports hybrid format: dictionary indices + full words for out-of-dict predictions.
 * 
 * Binary Format:
 * - Header: magic number + entry count
 * - Index section: sorted word indices with offsets (for binary search)
 * - Data section: prediction data with type flags (0=index, 1=full_word)
 */
class WordPredictionEngine {
    
    // MARK: - Properties
    
    /// Raw binary data from the predictions.bin file
    private let data: Data
    
    /// Direct pointer to memory for maximum speed
    private var basePointer: UnsafeRawPointer!
    
    /// Reference to the TrieEngine for word-index conversions
    private weak var trieEngine: TrieEngine?
    
    /// Constants
    private let headerSize = 8
    private let indexEntrySize = 6  // uint16 + uint32
    private let magicNumber: UInt32 = 0x50524544  // "PRED"
    
    // MARK: - Prediction Result
    
    struct Prediction {
        let word: String
        let probability: Float  // 0.0 to 1.0
    }
    
    // MARK: - Initialization
    
    /// Initialize with a specific language prediction file
    /// - Parameters:
    ///   - filename: Base filename (e.g., "he", "en", "ar")
    ///   - trieEngine: Associated TrieEngine for word lookups
    init?(filename: String, trieEngine: TrieEngine?) {
        // Try to load the prediction file
        guard let fileData = WordPredictionEngine.loadPredictionData(filename: filename) else {
            print("❌ WordPredictionEngine: Prediction file '\(filename)_predictions.bin' not found.")
            return nil
        }
        
        self.data = fileData
        self.trieEngine = trieEngine
        
        // Get a raw pointer to the bytes
        self.data.withUnsafeBytes { (rawBuffer: UnsafeRawBufferPointer) in
            self.basePointer = rawBuffer.baseAddress
        }
        
        // Verify magic number
        let magic = getUInt32(at: 0)
        guard magic == magicNumber else {
            print("❌ WordPredictionEngine: Invalid magic number: 0x\(String(magic, radix: 16))")
            return nil
        }
        
        let entryCount = getUInt32(at: 4)
        print("✅ WordPredictionEngine: Loaded '\(filename)_predictions' with \(entryCount) source words.")
    }
    
    /// Try to load prediction data from various locations
    private static func loadPredictionData(filename: String) -> Data? {
        let filenames = ["\(filename)_predictions"]
        
        let bundles: [Bundle] = [
            Bundle.main,
            Bundle(for: WordPredictionEngine.self)
        ]
        
        for bundle in bundles {
            for name in filenames {
                // Try direct resource lookup
                if let path = bundle.path(forResource: name, ofType: "bin"),
                   let data = try? Data(contentsOf: URL(fileURLWithPath: path)) {
                    print("📚 WordPredictionEngine: Found '\(name).bin' at \(path)")
                    return data
                }
                
                // Try dict/bin subdirectory
                if let path = bundle.path(forResource: name, ofType: "bin", inDirectory: "dict/bin"),
                   let data = try? Data(contentsOf: URL(fileURLWithPath: path)) {
                    print("📚 WordPredictionEngine: Found '\(name).bin' in dict/bin at \(path)")
                    return data
                }
                
                // Try bundle resources directory
                let bundleDir = bundle.bundlePath
                let directPath = "\(bundleDir)/\(name).bin"
                if FileManager.default.fileExists(atPath: directPath),
                   let data = try? Data(contentsOf: URL(fileURLWithPath: directPath)) {
                    print("📚 WordPredictionEngine: Found '\(name).bin' in bundle at \(directPath)")
                    return data
                }
            }
        }
        
        // Try app group container
        for name in filenames {
            if let groupContainerURL = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: "group.com.issie.IssieBoardNG") {
                let predPath = groupContainerURL.appendingPathComponent("\(name).bin")
                if FileManager.default.fileExists(atPath: predPath.path),
                   let data = try? Data(contentsOf: predPath) {
                    print("📚 WordPredictionEngine: Found '\(name).bin' in app group at \(predPath.path)")
                    return data
                }
            }
        }
        
        return nil
    }
    
    // MARK: - Public API
    
    /// Get word predictions after a given word
    /// - Parameters:
    ///   - word: The word that was just typed
    ///   - limit: Maximum number of predictions to return
    /// - Returns: Array of predictions sorted by probability
    func getPredictions(afterWord word: String, limit: Int = 6) -> [Prediction] {
        guard let trieEngine = trieEngine else {
            print("⚠️ WordPredictionEngine: No TrieEngine available")
            return []
        }
        
        // Get the word's index from the trie
        guard let wordIndex = trieEngine.getWordIndex(word.lowercased()) else {
            // Word not in dictionary - no predictions available
            return []
        }
        
        return getPredictions(forWordIndex: wordIndex, limit: limit)
    }
    
    /// Get predictions for a specific word index
    /// - Parameters:
    ///   - wordIndex: The index of the source word
    ///   - limit: Maximum number of predictions to return
    /// - Returns: Array of predictions
    func getPredictions(forWordIndex wordIndex: UInt16, limit: Int = 6) -> [Prediction] {
        // Binary search for the word index in the index section
        guard let dataOffset = findDataOffset(for: wordIndex) else {
            // No predictions for this word
            return []
        }
        
        // Parse predictions at the data offset
        return parsePredictions(at: dataOffset, limit: limit)
    }
    
    // MARK: - Binary Search
    
    /// Find the data offset for a given word index using binary search
    private func findDataOffset(for wordIndex: UInt16) -> Int? {
        let entryCount = Int(getUInt32(at: 4))
        var left = 0
        var right = entryCount - 1
        
        while left <= right {
            let mid = (left + right) / 2
            let entryOffset = headerSize + mid * indexEntrySize
            let midIndex = getUInt16(at: entryOffset)
            
            if midIndex == wordIndex {
                // Found it - return the data offset
                return Int(getUInt32(at: entryOffset + 2))
            } else if midIndex < wordIndex {
                left = mid + 1
            } else {
                right = mid - 1
            }
        }
        
        return nil
    }
    
    // MARK: - Prediction Parsing
    
    /// Parse prediction data at a specific offset
    private func parsePredictions(at offset: Int, limit: Int) -> [Prediction] {
        var predictions: [Prediction] = []
        var currentOffset = offset
        
        // Read prediction count
        let count = getUInt8(at: currentOffset)
        currentOffset += 1
        
        let actualCount = min(Int(count), limit)
        
        // Parse each prediction
        for _ in 0..<actualCount {
            let type = getUInt8(at: currentOffset)
            let probability = Float(getUInt8(at: currentOffset + 1)) / 255.0
            currentOffset += 2
            
            let word: String?
            
            if type == 0 {
                // Type 0: Dictionary index reference
                let dictWordIndex = getUInt16(at: currentOffset)
                currentOffset += 2
                
                // Convert index to word using TrieEngine
                word = trieEngine?.getWord(atIndex: dictWordIndex)
                
            } else {
                // Type 1: Full word stored inline
                let length = Int(getUInt8(at: currentOffset))
                currentOffset += 1
                
                // Read UTF-8 bytes
                let bytes = data.subdata(in: currentOffset..<(currentOffset + length))
                word = String(data: bytes, encoding: .utf8)
                currentOffset += length
            }
            
            if let word = word {
                predictions.append(Prediction(word: word, probability: probability))
            }
        }
        
        return predictions
    }
    
    // MARK: - Low-Level Memory Access
    
    private func getUInt8(at offset: Int) -> UInt8 {
        return basePointer.advanced(by: offset).load(as: UInt8.self)
    }
    
    private func getUInt16(at offset: Int) -> UInt16 {
        // Use loadUnaligned to handle potentially unaligned access
        return basePointer.advanced(by: offset).loadUnaligned(as: UInt16.self)
    }
    
    private func getUInt32(at offset: Int) -> UInt32 {
        // Use loadUnaligned to handle potentially unaligned access
        return basePointer.advanced(by: offset).loadUnaligned(as: UInt32.self)
    }
}

File: android/app/src/main/java/com/issieboardng/shared/WordPredictionEngine.kt
package com.issieboardng.shared

import android.content.Context
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * WordPredictionEngine - Next-word prediction engine
 * 
 * Port of ios/Shared/WordPredictionEngine.swift
 * 
 * Loads binary prediction files and provides next-word predictions based on bigram probabilities.
 * Uses hybrid storage format (dictionary indices + inline words) for space efficiency.
 * 
 * Binary format:
 * - Header: magic number (4 bytes) + entry count (4 bytes)
 * - Index: sorted array of (word_index: UInt16, offset: UInt32) pairs
 * - Data: prediction entries with probabilities and target words
 */
class WordPredictionEngine private constructor(
    private val data: ByteBuffer,
    private val entryCount: Int,
    private val trieEngine: TrieEngine?
) {
    
    companion object {
        private const val MAGIC_NUMBER = 0x50524544 // "PRED" in ASCII
        
        /**
         * Create a WordPredictionEngine by loading from assets
         * @param context Android context for asset access
         * @param filename Prediction file name (e.g., "he_predictions.bin")
         * @param trieEngine Optional TrieEngine for index-to-word conversion
         * @return WordPredictionEngine instance or null if loading fails
         */
        fun create(context: Context, filename: String, trieEngine: TrieEngine?): WordPredictionEngine? {
            val data = loadPredictionData(context, filename) ?: return null
            
            // Validate magic number
            val magic = data.getInt(0)
            if (magic != MAGIC_NUMBER) {
                debugLog("❌ WordPredictionEngine: Invalid magic number in $filename")
                return null
            }
            
            val entryCount = data.getInt(4)
            debugLog("✅ WordPredictionEngine: Loaded $filename with $entryCount entries")
            
            return WordPredictionEngine(data, entryCount, trieEngine)
        }
        
        private fun loadPredictionData(context: Context, filename: String): ByteBuffer? {
            try {
                val inputStream = context.assets.open(filename)
                val bytes = inputStream.readBytes()
                inputStream.close()
                
                val buffer = ByteBuffer.wrap(bytes)
                buffer.order(ByteOrder.LITTLE_ENDIAN)
                
                return buffer
            } catch (e: IOException) {
                debugLog("❌ WordPredictionEngine: File '$filename' not found")
                return null
            }
        }
    }
    
    /**
     * Prediction result with word and probability
     */
    data class Prediction(
        val word: String,
        val probability: UByte
    )
    
    /**
     * Get next-word predictions for a given source word
     * @param afterWord The word to predict after
     * @param limit Maximum number of predictions to return (default: 6)
     * @return List of predictions sorted by probability (highest first)
     */
    fun getPredictions(afterWord: String, limit: Int = 6): List<Prediction> {
        // Get word index from trie
        val wordIndex = trieEngine?.getWordIndex(afterWord) ?: run {
            debugLog("🔮 WordPredictionEngine: Word '$afterWord' not found in trie")
            return emptyList()
        }
        
        // Binary search in index to find prediction data offset
        val offset = binarySearchIndex(wordIndex) ?: run {
            debugLog("🔮 WordPredictionEngine: No predictions for word '$afterWord' (index $wordIndex)")
            return emptyList()
        }
        
        // Read predictions from data section
        return readPredictions(offset, limit)
    }
    
    /**
     * Binary search the index section to find the offset for a word
     * Index section starts at byte 8 and contains (word_index: UInt16, offset: UInt32) pairs
     */
    private fun binarySearchIndex(wordIndex: UShort): Int? {
        var left = 0
        var right = entryCount - 1
        
        while (left <= right) {
            val mid = (left + right) / 2
            val entryAddress = 8 + (mid * 6) // Header=8, each entry=6 bytes
            
            val currentWordIndex = data.getShort(entryAddress).toUShort()
            
            when {
                currentWordIndex == wordIndex -> {
                    // Found it - return the offset
                    val offset = data.getInt(entryAddress + 2)
                    return offset
                }
                currentWordIndex < wordIndex -> left = mid + 1
                else -> right = mid - 1
            }
        }
        
        return null
    }
    
    /**
     * Read prediction data from the data section
     * Format: prediction_count (UInt8) followed by predictions
     * Each prediction: type (UInt8), probability (UInt8), then either:
     *   - type=0: word_index (UInt16)
     *   - type=1: length (UInt8), utf8_bytes
     */
    private fun readPredictions(offset: Int, limit: Int): List<Prediction> {
        val predictions = mutableListOf<Prediction>()
        var position = offset
        
        // Read prediction count
        val count = data.get(position).toUByte()
        position += 1
        
        val actualCount = minOf(count.toInt(), limit)
        
        for (i in 0 until actualCount) {
            val type = data.get(position).toUByte()
            position += 1
            
            val probability = data.get(position).toUByte()
            position += 1
            
            val word = when (type.toInt()) {
                0 -> {
                    // Type 0: Index reference
                    val wordIndex = data.getShort(position).toUShort()
                    position += 2
                    
                    trieEngine?.getWord(wordIndex) ?: run {
                        debugLog("⚠️ WordPredictionEngine: Could not resolve index $wordIndex")
                        continue
                    }
                }
                1 -> {
                    // Type 1: Inline word
                    val length = data.get(position).toUByte().toInt()
                    position += 1
                    
                    val bytes = ByteArray(length)
                    for (j in 0 until length) {
                        bytes[j] = data.get(position + j)
                    }
                    position += length
                    
                    String(bytes, Charsets.UTF_8)
                }
                else -> {
                    debugLog("⚠️ WordPredictionEngine: Unknown type $type")
                    continue
                }
            }
            
            predictions.add(Prediction(word, probability))
        }
        
        // Sort by probability (highest first)
        predictions.sortByDescending { it.probability }
        
        return predictions
    }
}